pallets/frequency-tx-payment/src/lib.rs

```diff
@@ -0,0 +1,428 @@
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config + pallet_transaction_payment::Config + pallet_msa::Config
+	{
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The overarching call type.
+		type RuntimeCall: Parameter
+			+ Dispatchable<RuntimeOrigin = Self::RuntimeOrigin, PostInfo = PostDispatchInfo>
+			+ GetDispatchInfo
+			+ From<frame_system::Call<Self>>
+			+ IsSubType<Call<Self>>
+			+ IsType<<Self as frame_system::Config>::RuntimeCall>;
+
+		/// The type that replenishes and keeps capacity balances.
+		type Capacity: Replenishable + Nontransferable;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// The type that checks what transactions are capacity with their stable weights.
+		type CapacityCalls: GetStableWeight<<Self as Config>::RuntimeCall, Weight>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Dispatch the given call as a sub_type of pay_with_capacity. Calls dispatched in this
+		/// fashion, if allowed, will pay with Capacity,
+		#[pallet::call_index(0)]
+		#[pallet::weight({
+		let dispatch_info = call.get_dispatch_info();
+		(< T as Config >::WeightInfo::pay_with_capacity().saturating_add(dispatch_info.weight), dispatch_info.class)
+		})]
+		pub fn pay_with_capacity(
+			origin: OriginFor<T>,
+			call: Box<<T as Config>::RuntimeCall>,
+		) -> DispatchResult {
+			ensure_signed(origin.clone())?;
+
+			let _result = call.dispatch(origin);
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn compute_capacity_fee(
+		len: u32,
+		class: DispatchClass,
+		extrinsic_weight: Weight,
+	) -> BalanceOf<T> {
+		let weight_fee = Self::weight_to_fee(extrinsic_weight);
+
+		let len_fee = Self::length_to_fee(len);
+		let base_fee = Self::weight_to_fee(T::BlockWeights::get().get(class).base_extrinsic);
+
+		let fee = base_fee.saturating_add(weight_fee).saturating_add(len_fee);
+		fee
+	}
+
+	/// Compute the length portion of a fee by invoking the configured `LengthToFee` impl.
+	pub fn length_to_fee(length: u32) -> BalanceOf<T> {
+		T::LengthToFee::weight_to_fee(&Weight::from_ref_time(length as u64))
+	}
+
+	/// Compute the unadjusted portion of the weight fee by invoking the configured `WeightToFee`
+	/// impl. Note that the input `weight` is capped by the maximum block weight before computation.
+	pub fn weight_to_fee(weight: Weight) -> BalanceOf<T> {
+		// cap the weight to the maximum defined in runtime, otherwise it will be the
+		// `Bounded` maximum of its data type, which is not desired.
+		let capped_weight = weight.min(T::BlockWeights::get().max_block);
+		T::WeightToFee::weight_to_fee(&capped_weight)
+	}
+}
+
+/// Custom Transaction Validity Errors for ChargeFrqTransactionPayment
```

Comment:
nit: From here down appears to be related to the Signed Extension--for readability purposes, consider moving to a separate file, as this file gets a bit large to parse through otherwise.

Also update the comments copied from the boilerplate Signed Extension--for instance, we don't do tipping with Capacity, but the comments still reference it.